Parse getStatus as JSON not XML

ckelleyRH opened the pull request:
On `dogtagpki/pki` master XML is being replaced by JSON, getStatus will return JSON in PKI 11.0+.

The PR for `dogtagpki/pki` that makes this change necessary is: https://github.com/dogtagpki/pki/pull/3674

To test, I forked the nightly copr build of `freeipa` and added in this patch. I then ran the `dogtagpki/pki` CI against my copr fork instead of the nightly freeipa copr. `dogtagpki/pki` CI fails against the nightly `freeipa` build, but succeeds with my patched build.

This is my first PR to this repo (hello everyone!) so if I need to do anything in addition, like bump a version or document that there is a breaking API change, then I would be grateful if someone could point me in the right direction!

rcritten commented:
/azp run

rcritten commented on install/tools/ipa-pki-wait-running.in:
```diff
@@ -74,9 +74,13 @@ def get_status(conn, timeout):
+    try:
+        json_response = json.loads(response)
+        status = json_response['Response']['Status']
+    except Exception as e:
```
Can a more specific exception be caught?

ckelleyRH replied:
Please see reply to bottom comment.

rcritten commented on install/tools/ipa-pki-wait-running.in:
```diff
@@ -74,9 +74,13 @@ def get_status(conn, timeout):
+        status = json_response['Response']['Status']
+    except Exception as e:
+        error = e.message
     logging.debug("Got status '%s', error '%s'", status, error)
```
Please use logger.debug()

ckelleyRH replied:
Done in next revision.

rcritten commented on install/tools/ipa-pki-wait-running.in:
```diff
@@ -74,9 +74,13 @@ def get_status(conn, timeout):
+    error = None
+    try:
+        json_response = json.loads(response)
+        status = json_response['Response']['Status']
```
What happens if valid JSON is returned but these elements are not in it?

ckelleyRH replied:
`status` is `None` going into this block, so line 81 will `raise KeyError` and `status` will remain `None`.

I tried to adhere to the original intent of the code so as to not change behaviour. `status = root.findtext("Status")` returns `None` if the string is not found, so I made the new code do the same. Likewise for the error.

I can certainly catch more specific exceptions, `JSONDecodeError` from line 80 and `KeyError` for line 81. Done in next revision.

rcritten commented:
Can you update the commit message to include a link to the dogtag change so we have more context in the future?

Is this compatible with current dogtag or does this need to be pinned to a specific release?

ckelleyRH commented:
@rcritten I updated the commit message to include the dogtag PR link as requested. This change will be live from the next release of dogtag, it is not being backported.

ckelleyRH commented:
I think I fixed the linter issues, now it appears the CI is failing when starting a CA instance - but I haven't worked out why yet. The CA should be reporting its status as XML, which should cause the new code here to return a non-`None` error and a `None` status, but something is not right.

ckelleyRH commented:
Actually, now I think about it some more, I think this is the expected behaviour now. When I made my change in `dogtagpki/pki`, the IPA tests failed in a similar way to this - getStatus not returning a status causes everything to grind to a halt.

@rcritten what's the best thing to do here? I would think something similar to what I did in the other repo - provide a build of dogtag with my change in so that getStatus returns JSON?

rcritten commented:
There is no generic way to inject a new external build into the IPA CI. I'm looking into the possibilities without having to create a new vagrant image.

rcritten commented:
@ckelleyRH what is a pki repo we can test against?

I wonder if it makes sense to determine the type of data returned and parse as appropriate so that IPA can operate against new and old releases. It probably isn't a ton of code and will make integration easier.

ckelleyRH commented:
@rcritten I have modified the commit so that if parsing JSON fails then it tries the original XML code, as discussed. I have added an additional debug log message so that we can tell whether we went down the JSON path or the XML path. In addition, [here](https://copr.fedorainfracloud.org/coprs/ckelley/pki/) is a copr repo with a PKI build that includes my patch from `dogtagpki/pki`. Thanks

rcritten commented:
OK I'll work on a way to inject the copr repo into the images so we can test with the updated packages.

rcritten commented:
Kicking off full CI

ckelleyRH commented:
@rcritten Looks like CI issue rather than issue with the patch.

fcami commented:
@ckelleyRH yes indeed.

fcami commented:
Issue => https://pagure.io/freeipa/issue/8944
PR => https://github.com/freeipa/freeipa/pull/5953
Thanks for the heads-up @ckelleyRH

rcritten commented:
I was able to show in the other PR that this patch will interoperate just fine with update CA code.

@ckelleyRH can you add a standalone patch to set the default log level to DEBUG in ipa-pki-wait-running, similar to my patch in the other PR,  so we can see the message in the journal? It seems obvious that the fallback is working since the PR is passing but I'd like to document the message.

This will have the side effect of letting us see all the messages to determine if the tool is spamming the journal. There are other DEBUG messages which can never fire (because hardcoded debug level) and other dubious info vs warning messages. If the output isn't too terrible perhaps we can clean this up at the same time.

ckelleyRH commented:
@rcritten I added in a new commit that lowers the logger threshold. Hopefully the CI doesn't just die, we are experiencing CI pain with Github Actions.

rcritten commented:
From http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/3d5926d6-f606-11eb-ad97-5254008d789d/test_integration-test_installation.py-TestInstallMaster-test_install_master/master.ipa.test/journal.gz

Aug 05 16:08:45 master.ipa.test ipa-pki-wait-running[20155]: urllib3.connectionpool: Starting new HTTP connection (14): master.ipa.test:8080
Aug 05 16:08:45 master.ipa.test ipa-pki-wait-running[20155]: urllib3.connectionpool: http://master.ipa.test:8080 "GET /ca/admin/ca/getStatus HTTP/1.1" 200 169
Aug 05 16:08:45 master.ipa.test ipa-pki-wait-running[20155]: ipa-pki-wait-running: Response is not valid JSON, try XML
Aug 05 16:08:45 master.ipa.test ipa-pki-wait-running[20155]: ipa-pki-wait-running: Got status 'running', error 'None'
Aug 05 16:08:45 master.ipa.test ipa-pki-wait-running[20155]: ipa-pki-wait-running: Success, subsystem ca is running!

It falls back successfully so we know it works both with old and newer dogtag.

Ideally this tool would take a --debug option to bump the log level but I think its out of scope of this change.

@ckelleyRH can you drop the debug commit? Once CI is green again I'll merge. Thanks.

rcritten commented:
master:

* 9310366a6fe2523b660ccc4bc0dedb334e882723 Parse getStatus as JSON not XML
